packages/opencv/lib/index.js

```diff
@@ -0,0 +1,629 @@
+
+  const {
+    visualize = false,
+    threshold = DEFAULT_MATCH_THRESHOLD,
+    multiple = false,
+    matchNeighbourThreshold = MATCH_NEIGHBOUR_THRESHOLD,
+    method = DEFAULT_MATCHING_METHOD,
+  } = options;
+
+  let fullImg, partialImg, matched;
+
+  try {
+    ([fullImg, partialImg] = await B.all([
+      cvMatFromImage(fullImgData),
+      cvMatFromImage(partialImgData)
+    ]));
+    matched = new cv.Mat();
+    const results = [];
+    let visualization = null;
+
+    try {
+      cv.matchTemplate(fullImg, partialImg, matched, toMatchingMethod(method));
+      const minMax = cv.minMaxLoc(matched);
+
+      if (multiple) {
+        const matches = [];
+        for (let row = 0; row < matched.rows; row++) {
+          for (let col = 0; col < matched.cols; col++) {
+            const score = matched.floatAt(row, col);
+            if (score >= threshold) {
+              matches.push({score, x: col, y: row});
+            }
+          }
+        }
+
+        const nearMatches = filterNearMatches(matches, matchNeighbourThreshold);
+
+        for (const {x, y, score} of nearMatches) {
+          results.push({
+            score,
+            rect: {
+              x, y,
+              width: partialImg.cols,
+              height: partialImg.rows
+            }
+          });
+        }
+      } else if (minMax.maxVal >= threshold) {
+        const {x, y} = method.includes('SQDIFF') ? minMax.minLoc : minMax.maxLoc;
+        results.push({
+          score: minMax.maxVal,
+          rect: {
+            x, y,
+            width: partialImg.cols,
+            height: partialImg.rows
+          }
+        });
+      }
+
+      if (_.isEmpty(results)) {
+        // Below error message, `Cannot find any occurrences` is referenced in find by image
+        throw new Error(`Match threshold: ${threshold}. Highest match value ` +
+                        `found was ${minMax.maxVal}`);
+      }
+    } catch (e) {
+      // Below error message, `Cannot find any occurrences` is referenced in find by image
+      throw new Error(`Cannot find any occurrences of the partial image in the full image. ` +
+        `Original error: ${e.message}`);
+    }
+
+    if (visualize) {
+      const fullHighlightedImage = fullImg.clone();
+
+      for (const result of results) {
+        const singleHighlightedImage = fullImg.clone();
+
+        highlightRegion(singleHighlightedImage, result.rect);
+        highlightRegion(fullHighlightedImage, result.rect);
+        result.visualization = await jimpImgFromCvMat(singleHighlightedImage).getBufferAsync(Jimp.MIME_PNG);
+      }
+      visualization = await jimpImgFromCvMat(fullHighlightedImage).getBufferAsync(Jimp.MIME_PNG);
+    }
+    return {
+      rect: results[0].rect,
+      score: results[0].score,
+      visualization,
+      multiple: results
+    };
+  } finally {
+    try {
+      fullImg.delete(); partialImg.delete(); matched.delete();
+    } catch (ign) {}
+  }
+}
+
+function jimpImgFromCvMat (mat) {
```

Comment:
it would be nice to add docstrings to these new functions

Reply:
ok, added.